internal/service/resourcepolicy/model.go

```diff
@@ -0,0 +1,85 @@
+package resourcepolicy
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/conversion"
+	"go.mongodb.org/atlas-sdk/v20240805003/admin"
+)
+
+func NewTFResourcePolicyModel(ctx context.Context, input *admin.ApiAtlasResourcePolicy) (*TFResourcePolicyModel, diag.Diagnostics) {
+	createdByUser, diags := NewUserMetadataObjectType(ctx, input.CreatedByUser)
+	if diags.HasError() {
+		return nil, diags
+	}
+	lastUpdatedByUser, diags := NewUserMetadataObjectType(ctx, input.LastUpdatedByUser)
+	if diags.HasError() {
+		return nil, diags
+	}
+	policies, diags := NewPolicyObjectType(ctx, input.Policies)
+	if diags.HasError() {
+		return nil, diags
+	}
+	return &TFResourcePolicyModel{
+		CreatedByUser:     *createdByUser,
+		CreatedDate:       types.StringPointerValue(conversion.TimePtrToStringPtr(input.CreatedDate)),
+		ID:                types.StringPointerValue(input.Id),
+		LastUpdatedByUser: *lastUpdatedByUser,
+		LastUpdatedDate:   types.StringPointerValue(conversion.TimePtrToStringPtr(input.LastUpdatedDate)),
+		Name:              types.StringPointerValue(input.Name),
+		OrgID:             types.StringPointerValue(input.OrgId),
+		Policies:          *policies,
+		Version:           types.StringPointerValue(input.Version),
+	}, nil
+}
+
+func NewUserMetadataObjectType(ctx context.Context, input *admin.ApiAtlasUserMetadata) (*types.Object, diag.Diagnostics) {
```

Comment:
an alternative would be to have *diag.Diagnostics as param and add errors there

Reply:
curious to see how this function will be used

Reply:
It is used on line 13+17.
Why is it public you may ask?
1. Makes it easier to unit test
2. When generating the code, we don't know what will be usages. 
3. 🤔 Don't see why not?

WDYT?

Reply:
bfa7fa96fbac898a06cfc596d083d7cc75cc8ff1 @lantoli 
It looks cleaner now. Thank you for the suggestion ❤️